### DISCUSSION TITLE
On Sveltekit + Vercel: Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/node_modules/wavesurfer.js/dist/plugins/regions.esm.js'

### DISCUSSION THREAD

**Discussion by Artur-Galstyan** [katspaugh/wavesurfer.js#3066] [Q&A]
## Bug description
When you deploy an app with the Wavesurfer + Region plugin, you get a 500 error because:

```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/node_modules/wavesurfer.js/dist/plugins/regions.esm.js' imported from /var/task/.svelte-kit/output/server/entries/pages/_page.svelte.js
    at new NodeError (node:internal/errors:399:5)
    at finalizeResolution (node:internal/modules/esm/resolve:331:11)
    at moduleResolve (node:internal/modules/esm/resolve:994:10)
    at moduleResolveWithNodePath (node:internal/modules/esm/resolve:938:12)
    at defaultResolve (node:internal/modules/esm/resolve:1202:79)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```
However, locally, everything works, even if you build and preview the project.
```
npm run build && npm run preview
```
## Environment
 - Browser: Chrome, Firefox

## Minimal code snippet
```svelte
<script lang="ts">
    import { onMount } from "svelte";
    import WaveSurfer from "wavesurfer.js";
    import RegionsPlugin from "wavesurfer.js/dist/plugins/regions.js";
    import type { Region } from "wavesurfer.js/dist/plugins/regions.js";

    onMount(() => {
        const ws = WaveSurfer.create({
            container: "#waveform",
            waveColor: "rgb(200, 0, 200)",
            progressColor: "rgb(100, 0, 100)",
            url: "https://plazma-public-songs.s3.amazonaws.com/g989a7rqv6s7jlv.mp3"
        });
        const wsRegions = ws.registerPlugin(RegionsPlugin.create());

        // Give regions a random color when they are created
        const random = (min, max) => Math.random() * (max - min) + min;
        const randomColor = () =>
            `rgba(${random(0, 255)}, ${random(0, 255)}, ${random(
                0,
                255
            )}, 0.5)`;

        // Create some regions at specific time ranges
        ws.on("decode", () => {
            // Regions
            let region: Region = wsRegions.addRegion({
                start: 0,
                end: 8,
                content: "Resize me",
                color: randomColor(),
                drag: false,
                resize: true
            });
        });
    });
</script>

<div id="waveform" />
```

I've also taken the courtesy of creating an MVP project to see the issue:
https://github.com/Artur-Galstyan/wavesurfer-vercel-bug

I've deployed it here if you want to check:
https://wavesurfer-vercel-bug.vercel.app/

And thanks @katspaugh for all your help!

**Comment by katspaugh** [comment]
I’m not super knowledgeable in either Svelte or Vercel, but it sounds like it cannot load ESM modules in this setup.

You could try adding `“module”: “esnext”` to the tsconfig of that project.

I’m going to convert this to a Q&A discussion since it’s not clear yet if something needs to be done on wavesurfer’s side.

**Reply by Artur-Galstyan** [reply]
It seems both Netlify as well as Vercel have this problem and this only happens with plugins if you import like so

```typescript
import RegionsPlugin from "wavesurfer.js/dist/plugins/regions.js";
import type { Region } from "wavesurfer.js/dist/plugins/regions.js";
import TimelinePlugin from "wavesurfer.js/dist/plugins/timeline.js";
```

If you just import wavesurfer, everything is fine

```
import WaveSurfer from "wavesurfer.js";
```

I'll open another issue on Sveltekit. This is actually quite problematic for me as these plugins are very important :D

**Reply by Artur-Galstyan** [reply]
Just FYI, I created an [issue at sveltekit](https://github.com/sveltejs/kit/issues/10468). Let's see what happens :)

**Reply by Artur-Galstyan** [reply]
I've found out something interesting:

It works up until the release of [7.0.0-beta.15](https://github.com/katspaugh/wavesurfer.js/releases/tag/7.0.0-beta.15). After `7.0.0.-beta.16` it stops working (I went through all the releases lol). Something between those two releases happened, which is causing this issue.

**Reply by katspaugh** [reply]
Yes, beta.15 used to [export compiled files from TypeScript](https://github.com/katspaugh/wavesurfer.js/blob/7.0.0-beta.15/package.json#L24-L42) directly, while beta.16 switched to [rollup-built single-file ESM modules](https://github.com/katspaugh/wavesurfer.js/blob/7.0.0-beta.16/package.json#L24-L42).

Both are ESM modules though, so I'm not sure why beta.15 works and 16 doesn't.

**Reply by katspaugh** [reply]
The file itself is definitely there, and it's a valid module. Maybe it's not uploaded to Netlify/Vercel because it's called `regions.esm.js` and not `regions.js`?

```
~/S/wavesurfer-vercel-bug (main|✔) $ ls node_modules/wavesurfer.js/dist/plugins/regions.esm.js
node_modules/wavesurfer.js/dist/plugins/regions.esm.js
~/S/wavesurfer-vercel-bug (main|✔) $ node
Welcome to Node.js v18.16.0.
Type ".help" for more information.
> const Regions = import('wavesurfer.js/dist/plugins/regions.js')
undefined
> Regions
Promise {
  [Module: null prototype] { default: [class s extends e] },
  [Symbol(async_id_symbol)]: 187,
  [Symbol(trigger_async_id_symbol)]: 5
}
```

**Reply by katspaugh** [reply]
I would argue that's still a bug with Svelte's build system (also rollup AFAIK?) that doesn't package this file correctly despite the package.json mapping.

**Reply by Artur-Galstyan** [reply]
Yes, I saw that beta.16 used rollup; the strange part is that in node_modules the `regions.esm.js` definitely exists. I'll have to check why it's not uploaded to Vercel (i.e. wait on replies on my other GH issue). Until then, I'll unfortunately have to stick with 7.0.0-beta.16 :(

**Reply by katspaugh** [reply]
As a workaround, you could probably import the module via its full path, like this:

```js
import Regions from "../../node_modules/wavesurfer.js/dist/plugins/regions.esm.js"
```

Then svelte will likely know that this exact file needs to be included in the build.

**Comment by Artur-Galstyan** [comment]
Btw: I don't know why but version 7.1.4 works again. Sometimes, software is simply black voodoo magic 🤷

**Reply by katspaugh** [reply]
It's not magic, I've intentionally added .js files (in addition to .esm.js) to the package. 😄

**Comment by s-gbz** [comment]
Only semi-related, but for documentation's sake, I had the same issue with `ng build` in Angular.
Solution as provided by [Matt Saunders on SO](https://stackoverflow.com/a/56560971/11515036): 

Add the plugin path to the `script` section in `angular.json`.

```json
            "scripts": [
              "./node_modules/wavesurfer.js/dist/plugins/record.js"
            ]
```

Import it via

```js
import RecordPlugin from 'wavesurfer.js/dist/plugins/record.js';
```
